### DISCUSSION TITLE
Godot Logs

### DISCUSSION THREAD

**Discussion by emikrish** [sn-lab/MouseGoggles#14] [Q&A] [answered]
Hello!  I've put together the VR headset (thank you so so much for the help) and am trying to plan out how we might analyze behavior data using the Godot Logs. I've visually inspected some of the logs and it is a little bit hard to parse. The CSV has 4 columns, some of which I'm assuming are position data, and some of which are time/frame data?  I was wondering what exactly these default Godot logs are logging, and whether you recommend adding code to create custom logs? Thank you! :)

**Comment by misaacson01** [comment, marked as answer]
I'll be adding some instructions soon to the main page for reading these Godot logs, but to give you some info now:

Different experiments have a different logging structure, but there are commonalities. Every line (row) in the csv is some select data for one frame as it's rendered. Each column is a different datatype, where the last column is always the timestamp (in ms from experiment start) of that frame. Typically, position data make up many of the other columns, but it really depends on the experiment. I personally like this format since it's very complete, easy to analyze, and easy to adjust by simply adding or subtracting columns for whatever datatype you want to log. To see what is actually being stored in these logs, for each experiment, check out the "___Scene.gd" file relating to that experiment; you can either read it in Godot or just a text editor. There will be a line towards the end starting with "save_logs(...)" which lists the different variables stored in each column, in order. As for what gets stored: For a 2D navigation experiment, maybe you just want to store the mouse's position on the ground, in [x,z] coordinates (Godot treats the y-axis as the vertical axis), as well as the mouse's head angle. If you're measuring licking behavior, you want to log that. Maybe you want to log all the raw yaw/pitch/roll spherical treadmill values for more detailed offline analysis of movement. Or for the looming experiment, some variables about the current position of the looming object is logged. So you get the idea, it's really whatever you want to store from the experiment and it's fully customizable.

If you want to change what data is stored in these logs, there are 4 different areas of the "___Scene.gd" code that need to be edited:

1. In the `#logging` section towards the beginning of the script, initialize the blank vector for the data you want to log (e.g. `var yawLog := []`)
2. In the `#save data` section of `func _process(delta):` (the loop that runs every frame of the game), append the current value of the variable into the vector (e.g. `yawLog.append(head_yaw)`)
3. In the `next trial` section, find save_logs() and make sure your variable is in there, in the column order you want (e.g. `save_logs(current_rep,yawLog,...)`)
4. In `save_logs()` towards the very end of the script, make sure the number of columns here (both in the `save_logs()` function inputs and in `file.store_line()` inputs) matches the number of data columns you're logging. Not all inputs have to be logged as a column; for instance, I put the trial repetition number in the filename rather than in the file itself.

The current Godot project version on Github is 0.7 but I'm currently working with 0.9, and I'll upload a new project once it gets to 1.0 so there may be some small changes to the logging method when that happens. But feel free to read up on Godot programming and do things however you like, you're definitely not stuck with the data logging format this project currently uses!

**Reply by emikrish** [reply]
Thank you!! Is there a way to set the timestamp to start at the beginning of the actual scene (cliffscene, eg), rather than the "play button" time (sceneselect)?

**Reply by misaacson01** [reply]
Good suggestion, I'll try to change this for 1.0 as well. I imagine that moving the `var timestamp = ...` line to later in the script (e.g., to the end of `func _ready():`) might work?

**Comment by emikrish** [comment]
Hi Matthew!!  I hope you're doing well!   I am writing to double-check about the Godot Looming scenes -- the logs document certain "types" of looming stimuli (0, 1, 2, 3) -- what is the difference between these trial types?

Thank you!!
Emi

**Reply by misaacson01** [reply]
For that experiment, the conditions are:
0: looming object approaching from the left side
1: sky darkening on the left side, similar to the darkening that happens with condition 0 except that there is no looming motion
2: looming object approaching from the right side
3: sky darkening on the right side

The darkening conditions (1 and 3) are just sort of a negative control, so you might not need them. In newer versions of this experiments, I've been using 3 conditions: loom from the right, loom from the left, and loom from straight ahead. But you could also get away with just the 2 conditions (loom right/loom left)
